### ISSUE TITLE
Towards a very initial integration of computational storage features

### ISSUE THREAD

**Issue by pdamme** [daphne-eu/daphne#656]
@niclashedam and I have recently discussed how we could approach an *initial integration* of some first computational storage features into DAPHNE. *The focus is on where to add/change what in the DAPHNE code base.* For now, it's really just about an initial integration, there may be better ways to achieve it and we can surely refine and extend it in the future. Here is a summary:

### Idea/goals
- a C/C++ kernel that pushes certain pre-defined sequences/DAGs of operations down to computational storage
- should support an arbitrary number of inputs (matrices/frames/scalars) and outputs
- should be generic w.r.t. the concrete DAG of ops to push down, multiple such DAG/functions could be registered, they should be selected by an additional parameter
- should be usable from DaphneDSL
- question: how can we achieve this in DAPHNE?

### Disclaimer
- the goal is to give some pointers to (most of) the decisive points in the code base
- most of the following steps should be copy-paste-adapt of existing code snippets, at some points there will also be some additional work
- overall, I hope this will be helpful to get started

### Suggested steps
- **add new DaphneIR (MLIR) operation (say, `CompStorageOp`)**
  - see `src/ir/daphneir/DaphneOps.td`
    - this is where all DaphneIR operations (that the DAPHNE compiler can reason about) are defined in [MLIR’s Operation Definition Specification](https://mlir.llvm.org/docs/DefiningDialects/Operations/)
  - the new op should have variadic arguments and results, such that it can replace any DAG of operations
  - furthermore, the op should get the information for which function registered in computational storage it should execute (could be just an integer code or string name, to begin with)
  - take inspiration from `VectorizedPipelineOp` and `DistributedPipelineOp`, probably the most similar ones to your needs
- **add a new compiler pass that can replace a DAG of operations by a `CompStorageOp`**
  - new pass
    - take inspiration from existing lowering passes in `src/compiler/lowering/`
      - e.g., `RewriteSqlOpPass` is a relatively simple pass that replaces certain ops by other ops
    - for the identification of the operation DAGs, take inspiration from the [canonicalize methods](https://mlir.llvm.org/docs/Canonicalization/) in `src/ir/daphneir/DaphneDialect.cpp`
      - idea: to match A(B(C), D)
        - check if the given op is an A-op
        - if so
          - check if the first operand is a B-op, and if so, if its operand is a C-op
          - check if the second operand is a D-op
        - if that all holds, replace the A-op by your `CompStorageOp` (no need to erase any operations, they will be clearer away automatically later, if not needed elsewhere)
  - add pass to pass pipeline
    - the pass pipeline is defined in `src/compiler/execution/DaphneIrExecutor.cpp`
    - just before the vectorization pass (`createVectorizeComputationsPass`) could be a good place for the beginning (the integration with vectorization can be discussed later)
  - add config flags for turning the pass on/off and for printing the IR after the pass for debugging (optional)
    - several passes have such flags, take inspiration from, e.g., the vectorization pass
    - relevant for the config
      - definition: `src/parser/config/JsonParams.h`, `src/parser/config/ConfigParser.cpp`, `src/api/cli/DaphneUserConfig.h`, `UserConfig.json`
      - CLI args: `src/api/internal/daphne_internal.cpp`
      - usage (add the pass if flag in on, print the IR afterwards if flag is on): `src/compiler/execution/DaphneIrExecutor.cpp`
      - [partly documented](https://daphne-eu.github.io/daphne/Config/)
- **lowering `CompStorageOp` to a kernel call**
  - see `src/compiler/lowering/RewriteToCallKernelOpPass.cpp`
  - domain-specific DaphneIR ops are lowered to DaphneIR’s `CallKernelOp`, which still has the same arguments and results, but gets the name of the C function to call as the first argument, plus the `DaphneContext` as the last argument
    - those C functions can be found in `build/src/runtime/local/kernels/kernels.cpp` (generated as part of the DAPHNE build process)
    - you need to make sure that your kernel get pre-compiled, too (see below)
  - support your new `CompStorageOp` in `getNumODSOperands()` and `getODSOperandInfo()`, such that the pass knows how to correctly handle the variadic arguments of the op
  - supporting variadic results could require some extra work
    - either try it with just a single result first
    - or take inspiration from `DistributedPipelineKernelReplacement`
- **kernel for `CompStorageOp`**
  - write the kernel
    - interface/boilerplate
      - writing a kernel is [mostly documented](https://daphne-eu.github.io/daphne/development/ImplementBuiltinKernel/)
      - take inspiration from vectorized pipeline kernel: `src/runtime/local/kernels/VectorizedPipeline.h`
    - contents
      - your specific code to offload the computation to computational storage
  - register the kernel for pre-compilation
    - see `src/runtime/local/kernels/kernels.json`
      - take inspiration from `VectorizedPipeline`-kernel
    - this file contains the interfaces etc. of all kernels we want to pre-compile such that we can use them from the JIT-compiled DaphneDSL script
  - compilation chain
    - maybe you need to adapt the C++ compilation chain to link with any libs you require
    - see, e.g., `src/runtime/local/kernels/CMakeLists.txt`

**Comment by Nickromancer** [comment]
Hello Patrick, 
Thank you for the guide. We are working with daphne for our bachelor project and have some additional questions.
- Do we need interfaces for our CompStorageOps method like the VectorizedPipelineOp and DistributedPipelineOp methods have?
- Are there any resources/diagrams for the overall infrastructure of adding new features for daphneIR, lowering, passes, that would be relevant for us?
- Can you recommend any tutorials/documentation of technologies and languages used for the scope of the project?
Thanks in advance

**Comment by niclashedam** [comment]
Patrick, meet Nicholas (& Emilia), my two bachelor's students this semester. They will work on building the first computational storage capabilities of DAPHNE. We all appreciate your efforts to create this issue on GitHub and our discussions at HiPEAC in Munich.

Let me answer the questions from my perspective, and then Patrick can correct me if I need to be corrected and add any points I still need to include.

As for the first question, if I understand the two operations correctly, they inherit from `OpConversionPattern`, which rewrites the high-level operation to something more concrete. Your operation should use the same approach because the underlying operation will be more complicated and require some setup and teardown.

It makes sense for the second and third questions to understand what LLVM is and why it's interesting.

![image](https://github.com/daphne-eu/daphne/assets/770169/738a7226-6e82-4002-908f-f687438d12d3)
https://mukulrathi.com/create-your-own-programming-language/llvm-ir-cpp-api-tutorial/

LLVM takes some input language, in our case, DaphneDSL, transpiles it into DaphneIR (MLIR) and spits out an executable runtime plan. Recall the lexers and parsers from ITU's "Programs as Data" course. LLVM does all that but with optimisations and more.

**Comment by pdamme** [comment]
Hi @Nickromancer, Emilia, and @niclashedam! Sorry for the late response, and thanks for supporting DAPHNE.

Regarding your questions:
1. I assume by interfaces you mean MLIR operation interfaces like `VectorizableOpInterface` (`src/ir/daphneir/DaphneVectorizableOpInterface.td`), which is used by the `VectorizeComputationsPass` (`src/compiler/lowering/VectorizeComputationsPass.cpp`) DAPHNE compiler pass to identify DaphneIR operations that can be vectorized. For your purpose, I wouldn't recommend introducing a new interface to identify the operations you want to push down to computational storage. The reason is that, as far as I understand, you want to identify concrete DAGs of DaphneIR operations that are semantically equivalent to hand-written C/C++ functions that were compiled to eBPF code and registered with the device. An interface could only be attached to a class of DaphneIR operations (see the `DeclareOpInterfaceMethods<...>` in `src/ir/daphneir/DaphneOps.td` for examples), not to a DAG of operations. Furthermore, interfaces are designed to handle all MLIR operations possessing the interface in a unified way, independent of which concrete operation it is. But since you anyway need to know the concrete (DAG of) operation(s), I think an interface will not help you much. Rather than that, I recommend identifying the DAGs of operations to replace by the `CompStorageOp` the way I suggested in *"add a new compiler pass that can replace a DAG of operations by a `CompStorageOp`"* in my comment above.
2. There is currently no DAPHNE-specific documentation on adding new DaphneIR operations or DAPHNE compiler passes. To add a new DaphneIR operation, you essentially need to add a new `def ...` in `src/ir/daphneir/DaphneOps.td` akin to those already present. The language used for that purpose is MLIR's [Operation Definition Specification](https://mlir.llvm.org/docs/DefiningDialects/Operations/), if you want to read up on the details. For adding a compiler pass, I recommend reading the relevant MLIR documentation (see below) and taking inspiration from existing DAPHNE compiler passes (as suggested above).
3. The following general resources could be useful to you: [MLIR documentation](https://mlir.llvm.org/docs/) (concepts explained), [MLIR doxygen](https://mlir.llvm.org/doxygen/namespacemlir.html) (API documentation), [MLIR discourse](https://discourse.llvm.org/c/mlir/31) (maybe someone had the same general MLIR-related question, or you could ask the MLIR developers; of course not intended for DAPHNE-specific questions, those are better asked here). In particular, the following articles on MLIR (non-DAPHNE-specific) could be interesting for your project: [ODS](https://mlir.llvm.org/docs/DefiningDialects/Operations/), [canonicalization](https://mlir.llvm.org/docs/Canonicalization/), [pass infrastruction](https://mlir.llvm.org/docs/PassManagement/), [pattern rewriting](https://mlir.llvm.org/docs/PatternRewriter/), and [dialect conversion](https://mlir.llvm.org/docs/DialectConversion/). Besides that, the [MLIR tutorials](https://mlir.llvm.org/docs/Tutorials/) may not hurt (though not all of those will be relevant to your project, they could give you some more background on MLIR).

I'd also like to clarify the role of LLVM in DAPHNE a bit: The processing in DAPHNE starts with a DaphneDSL script (possibly generated by DaphneLib, DAPHNE's Python API). The DaphneDSL script is turned into DaphneIR (custom MLIR dialect) by our ANTLR-based parser. The initial DaphneIR representation of the program is optimized and lowered by standard and custom MLIR passes in the DAPHNE compiler. The last lowering step converts to MLIR's llvm dialect, which is then translated to LLVM IR, just-in-time compiled, and executed. That also means that DAPHNE developers rarely get in contact with LLVM directly (unless they are really deep down in the compiler stack, which is not the case for your project). So MLIR is relevant to you, while LLVM isn't really.

I hope this answer is helpful. Please don't hesitate to ask if you have specific questions.

**Comment by philipportner** [comment]
Hi @niclashedam ,

maybe I can help you with that.

`error: ‘helloWorldOp’ was not declared in this scope` , the name of the op should start with a capital `H`

`error: no matching function for call to ‘mlir::OpBuilder::create<<expression error> >()’`, here you need to look at the function you are trying to call, the [signature](https://github.com/llvm/llvm-project/blob/20d454c79bbca7822eee88d188afb7a8747dac58/mlir/include/mlir/IR/Builders.h#L459) of `mlir::OpBuilder::create` expects at least a location. Together with the correct capitalization it should look like this:

```diff
-    return builder.create<helloWorldOp>();
+    return builder.create<HelloWorldOp>(loc);
``` 

Once you want to actually pass arguments to an Operation you create you will want to have a look at the `OpTy::build` function that is being used [here](https://github.com/llvm/llvm-project/blob/20d454c79bbca7822eee88d188afb7a8747dac58/mlir/include/mlir/IR/Builders.h#L462) in the `mlir::OpBuilder::create` call. That you can find in `build/src/ir/daphneir/DaphneOps.h.inc`, there you will find a class called `HelloWorldOp` which as multiple overloaded `build` functions.
I recommend using a `clangd` LSP client in your text editor of choice. [This](https://clangd.llvm.org/installation.html#editor-plugins) should help you set that up, the compilation database is already generated and can be found in `build/compile_commands.json`. With that you will get diagnostics if you do not match a call to `builder.create<...` and you can simply `Go To Definition` to inspect the generated `build` functions for the operator you are trying to create.


```
/home/dawn/daphne_dawn/src/parser/daphnedsl/DaphneDSLBuiltins.cpp:1221:40: error: template argument 1 is invalid
 1221 |     return builder.create<helloWorldOp>();
      |            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~
```

As you did not define a template function, you cannot instantiate an [abbreviated function template](https://en.cppreference.com/w/cpp/language/function_template).
Additionally, all our kernels expect a `DaphneContext` argument as last function argument.

```diff
-template<>
-void printHelloWorld() {
+void helloWorld(DCTX(ctx)) {
```

Lastly, you will have to fix the kernel information in `kernels.json`. The documentation for that is WIP but for your dummy function you have to add at least a single instantiation to the list of instantiations. Otherwise, the code that generates the wrapper functions that you find after compilation in `build/src/runtime/local/kernels/kernels.cpp` will not be created.

```diff
-           "instantiations": []
+           "instantiations": [[]]
```

Once that is fixed you will find the following function in `build/src/runtime/local/kernels/kernels.cpp`:
```
// ----------------------------------------------------------------------------
// helloWorld
// ----------------------------------------------------------------------------
void _helloWorld(DCTX(ctx)) {
        helloWorld(ctx);
}
```
This is the automatically generated wrapper that calls your kernel in `src/runtime/local/kernels/`.


I tested these patches in your branch and with them you can compile and run your test script which will call your new precompiled kernel and print `Hello World\n`.

> I would really appreciate if you could clarify this and also how to extend kernel calls to the DaphneDSL, this information was unclear in the Daphne documentation.

Once you want to add arguments to your function you will have to adapt the definition in `DaphneOps.td`, `kernels/HelloWorld.h` and `kernels.json`. You also need to pass the arguments when creating the operation in `DaphneDSLBuiltins.cpp` when calling the `mlir::OpBuilder::create` function.

BR,
Philipp

**Comment by Nickromancer** [comment]
Hi again!

I'm trying to add another kernel ```kernels/AddTwo.h``` that take two value types, add them together and returns the result with the same type. My goal with this is to understand the input and output configurations of kernels, also in regards to variadic data. The project builds just fine but when I try to call the function with e.g ```addTwo(2,2);```, it returns an error with: ```[error]: Parser error: unknown built-in function: 'addTwo'```.

I followed the same general steps as before with ```kernels/HelloWorld.h``` but I have no idea where the fault could possibly be. 

The project branch is this one: https://github.com/Nickromancer/daphne_dawn_nick/tree/Addition
Thanks in advance!

BR,
Nicholas

**Comment by philipportner** [comment]
> Hi again!
> 
> I'm trying to add another kernel `kernels/AddTwo.h` that take two value types, add them together and returns the result with the same type. My goal with this is to understand the input and output configurations of kernels, also in regards to variadic data. The project builds just fine but when I try to call the function with e.g `addTwo(2,2);`, it returns an error with: `[error]: Parser error: unknown built-in function: 'addTwo'`.
> 
> I followed the same general steps as before with `kernels/HelloWorld.h` but I have no idea where the fault could possibly be.
> 
> The project branch is this one: https://github.com/Nickromancer/daphne_dawn_nick/tree/Addition Thanks in advance!
> 
> BR, Nicholas

You have an unconditional `throw` before checking for your new built-in function.

https://github.com/Nickromancer/daphne_dawn_nick/blob/635ae9ea1267d49b570002082c07e61b4c829900/src/parser/daphnedsl/DaphneDSLBuiltins.cpp#L1225

**Comment by Nickromancer** [comment]
Hello,

My kernel ```kernels/AddTwo.h``` is now recognized by DaphneDSL, but when I try to call it, it returns with the error: 
```[error]: Parser error: something was expected to be an mlir::Value, but it was none```.

I've removed the variadic type from its dialect in ```Daphneir/DaphneOps.td``` to simplify the kernel and followed some of the elements in my implementation as in ```def Daphne_EwBinaryOp```.

The project: https://github.com/Nickromancer/daphne_dawn_nick/tree/Addition

Thank you for bearing with my rookie mistakes so far, I'm really grateful for the advice.

**Comment by philipportner** [comment]
> My kernel `kernels/AddTwo.h` is now recognized by DaphneDSL, but when I try to call it, it returns with the error: `[error]: Parser error: something was expected to be an mlir::Value, but it was none`.

While the error message could be improved (WIP), only two places in the code can produce such an error message, `src/parser/ParserUtils.h::198` and `src/parser/ParserUtils.h::204`, both of which throw in case the argument is not a `mlir::Value`. This gives you a strong hint and helps you with debugging your code.

`mlir::OpBuilder::create` returns a value of the type `OpTy` you are passing in as template parameter, in your case `AddTwoOp`. This type is not implicitly convertible to `mlir::Value`. The IR structure [documentation](https://mlir.llvm.org/docs/Tutorials/UnderstandingTheIRStructure) may help you.
Simply put, you are missing either a `static_cast` or a call on the operation like `.getRes()`.

**Comment by niclashedam** [comment]
Another question for you, @pdamme and @philipportner.
We would need to precompile a set of static eBPF programs for offload to computational storage. This is quite simple:

`clang -target bpf -c -Wno-implicit-function-declaration -fno-builtin -O3 bpf.c`

This will yield a `.o` file that we should be able to read in DAPHNE with the intent of transferring it to computational storage.
See section 3.2 in [this paper](https://hed.am/papers/2021-EBPF.pdf) for more info.

It will also be necessary to include `io_uring` in the build, so I suggest we place this behind a compilation flag. How do you think we should proceed?

**Comment by Nickromancer** [comment]
Hello @pdamme and @philipportner.
I want to make a kernel that initializes the global states for ```CompStorage``` in a kernel in Daphne. Where and how would I be able to access, create and maintain global states in Daphne?

**Comment by philipportner** [comment]
> Another question for you, @pdamme and @philipportner. We would need to precompile a set of static eBPF programs for offload to computational storage. This is quite simple:
> 
> `clang -target bpf -c -Wno-implicit-function-declaration -fno-builtin -O3 bpf.c`
> 
> This will yield a `.o` file that we should be able to read in DAPHNE with the intent of transferring it to computational storage. See section 3.2 in [this paper](https://hed.am/papers/2021-EBPF.pdf) for more info.
> 
> It will also be necessary to include `io_uring` in the build, so I suggest we place this behind a compilation flag. How do you think we should proceed?

Do you want to add [axboe/liburing](https://github.com/axboe/liburing) as a dependency? If so, you can do it similarly to other dependencies in DAPHNE like CUDA or Abseil. Check the `build.sh` script, maybe add a flag similar to `--cuda`?
To create the `.so` for your eBPF program, the compilation of that should be added to the CMake project, placing the `.so` file  in `lib/`.

Once everything is compiled, you'll need to add the `.so` to our `mlir::ExecutionEngine`, similar to what we already do [here](https://github.com/daphne-eu/daphne/blob/5de5748659d21f1e5b63c3d67d150b9cedd68b50/src/compiler/execution/DaphneIrExecutor.cpp#L252) .

**Comment by philipportner** [comment]
> Hello @pdamme and @philipportner. I want to make a kernel that initializes the global states for `CompStorage` in a kernel in Daphne. Where and how would I be able to access, create and maintain global states in Daphne?

You probably want to add whatever you need to the `DaphneContext` (as described [here](https://github.com/daphne-eu/daphne/blob/5de5748659d21f1e5b63c3d67d150b9cedd68b50/src/runtime/local/context/DaphneContext.h#L45-L53)).

**Comment by niclashedam** [comment]
> > Another question for you, @pdamme and @philipportner. We would need to precompile a set of static eBPF programs for offload to computational storage. This is quite simple:
> > `clang -target bpf -c -Wno-implicit-function-declaration -fno-builtin -O3 bpf.c`
> > This will yield a `.o` file that we should be able to read in DAPHNE with the intent of transferring it to computational storage. See section 3.2 in [this paper](https://hed.am/papers/2021-EBPF.pdf) for more info.
> > It will also be necessary to include `io_uring` in the build, so I suggest we place this behind a compilation flag. How do you think we should proceed?
> 
> Do you want to add [axboe/liburing](https://github.com/axboe/liburing) as a dependency? If so, you can do it similarly to other dependencies in DAPHNE like CUDA or Abseil. Check the `build.sh` script, maybe add a flag similar to `--cuda`? To create the `.so` for your eBPF program, the compilation of that should be added to the CMake project, placing the `.so` file in `lib/`.
> 
> Once everything is compiled, you'll need to add the `.so` to our `mlir::ExecutionEngine`, similar to what we already do [here](https://github.com/daphne-eu/daphne/blob/5de5748659d21f1e5b63c3d67d150b9cedd68b50/src/compiler/execution/DaphneIrExecutor.cpp#L252) .

Hi again. Thanks for the answer. The compilation yields an `.o` file -- not an `.so` file. And it shouldn't be linked into DAPHNE since it's compiled for computational storage. We just need a place to place this and compile it in isolation from the rest of DAPHNE, but within the build script.

And yes, it's axboe's library :-)

**Comment by Nickromancer** [comment]
Hi again @pdamme and @philipportner. 
I'm trying to extend the ```DaphneContext``` with a ```DelilahContext``` and I've followed the same structure as for the ```FPGAContext```. When trying to call the getter for Delilah from the DaphneContext I get the following error: 
```
In file included from /home/dawn/daphne_dawn/build/src/runtime/local/kernels/kernels.cpp:20:
/home/dawn/daphne_dawn/src/runtime/local/kernels/Delilah_init.h: In function ‘int delilahInit(DaphneContext*)’:
/home/dawn/daphne_dawn/src/runtime/local/kernels/Delilah_init.h:22:32: error: ‘struct DaphneContext’ has no member named ‘getDelilahContext’
   22 |     auto DelilahContext = ctx->getDelilahContext(0);
```

I'm also curious how keep my kernels under the ```kernels/DELILAH``` folder working properly with the include paths for ```kernels.cpp``` since its auto generated. I'm unsure how to check if my context is properly implemented, so I want to know if there are certain files in the project I have to keep a eye out for.

Thank you very much!

**Comment by Nickromancer** [comment]
Hello @pdamme and @philipportner!
When building my kernel ```Delilah_init``` I get this error:
```
ld.lld: error: undefined symbol: io_uring_queue_init
>>> referenced by kernels.cpp
>>>               src/runtime/local/kernels/CMakeFiles/AllKernels.dir/kernels.cpp.o:(delilahInit(DaphneContext*))
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
```
I'm trying to use ```<Liburing.h>``` for this kernel but it does not seem fully functional.
@niclashedam and I tried looking at ```/CMAKELists.txt```, where we think this line is needed but still does not solve the problem:
```find_package(liburing REQUIRED)```
We are still not entirely sure what is required to add this library properly into our kernels.
Thanks in advance! 

_Same working branch as before_

**Comment by philipportner** [comment]
> Hello @pdamme and @philipportner! When building my kernel `Delilah_init` I get this error:
> 
> ```
> ld.lld: error: undefined symbol: io_uring_queue_init
> >>> referenced by kernels.cpp
> >>>               src/runtime/local/kernels/CMakeFiles/AllKernels.dir/kernels.cpp.o:(delilahInit(DaphneContext*))
> collect2: error: ld returned 1 exit status
> ninja: build stopped: subcommand failed.
> ```
> 
> I'm trying to use `<Liburing.h>` for this kernel but it does not seem fully functional. @niclashedam and I tried looking at `/CMAKELists.txt`, where we think this line is needed but still does not solve the problem: `find_package(liburing REQUIRED)` We are still not entirely sure what is required to add this library properly into our kernels. Thanks in advance!
> 
> _Same working branch as before_

Hey, I can't find the modified `CMakeLists.txt` you are working with in your branch, but I would assume that you are not actually linking the `liburing` library. Similar to how it's done with `HW_LOC` and `PAPI` [here](https://github.com/Nickromancer/daphne_dawn_nick/blob/3fc3654241e05e72d4378c31cc4f4021088c28a8/src/runtime/local/kernels/CMakeLists.txt#L117-L120).

BR,
Philipp

**Comment by Nickromancer** [comment]
Hey again @philipportner @pdamme!

**Comment by philipportner** [comment]
> Hey again @philipportner @pdamme! I'm having difficulties when calling my `ctx->getDelilahContext(0);`, causing this error to occour:
> 
> ```
> dawn@nhed-2:~/daphne_dawn$ bin/daphne --delilah scripts/examples/Delilah.daph 
> bin/daphne(+0x14e50e2)[0x5649c56570e2]
> /lib/x86_64-linux-gnu/libc.so.6(+0x42520)[0x7ffa80842520]
> /home/dawn/daphne_dawn/bin/../lib/libAllKernels.so(_delilahInit__char+0x18)[0x7ffa7cad84b8]
> [0x7ffa81104060]
> [0x7ffa8110408d]
> [0x7ffa811040ad]
> bin/daphne(+0x1d0100b)[0x5649c5e7300b]
> bin/daphne(+0x1513dc0)[0x5649c5685dc0]
> bin/daphne(+0x151834d)[0x5649c568a34d]
> /lib/x86_64-linux-gnu/libc.so.6(+0x29d90)[0x7ffa80829d90]
> /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0x80)[0x7ffa80829e40]
> bin/daphne(+0x14e3be5)[0x5649c5655be5]
> [error]: Got an abort signal from the execution engine. Most likely an exception in a shared library. Check logs!
> [error]: Execution error: Returning from signal 11
> ```
> 
> Working branch: https://github.com/Nickromancer/daphne_dawn_nick/tree/Delilah_init

Hi, seems like you are missing the implementation for your `DelilahInitPass` [here](https://github.com/Nickromancer/daphne_dawn_nick/blob/Delilah_init/src/compiler/lowering/DelilahInitPass.cpp). Without the `CreateDelilahContextOp` your `createDELILAHContext` kernel is never called and thus the `delilah_contexts` of the `DaphneContext` is empty resulting in the out-of-bounds access when using `ctx->getDelilahContext(0);`.

I see you tried to do that in the [`InsertDaphneContextPass.cpp`](https://github.com/Nickromancer/daphne_dawn_nick/blob/d5a02bc70ef122329d52e6db2bfc58fad5419cc9/src/compiler/lowering/InsertDaphneContextPass.cpp#L71-L75) , but I think that's not working since you have a wrong condition in the `if` surrounding the creation of the `CreateDelilahContextOp`.

To make sure the correct operations are added in the pass pipeline it is useful to either use existing locations that call the `PrintIrPass` or to add your own. After applying the following patch you can see the IR directly after the `InsertDaphneContextPass` and figure if something is missing or not. You should be able to see the `CreateDelilahContextOp` in the IR.


```diff
diff --git a/src/compiler/execution/DaphneIrExecutor.cpp b/src/compiler/execution/DaphneIrExecutor.cpp
index 37666f9..32e6e6e 100644
--- a/src/compiler/execution/DaphneIrExecutor.cpp
+++ b/src/compiler/execution/DaphneIrExecutor.cpp
@@ -190,6 +190,9 @@ bool DaphneIrExecutor::runPasses(mlir::ModuleOp module) {
     pm.addNestedPass<mlir::func::FuncOp>(
         mlir::daphne::createInsertDaphneContextPass(userConfig_));

+    pm.addPass(mlir::daphne::createPrintIRPass(
+        "IR after inserting DaphneContext (and others depending on settings.):"));
+
 #ifdef USE_CUDA
     if (userConfig_.use_cuda)
         pm.addNestedPass<mlir::func::FuncOp>(

```

BR,
Philipp

**Comment by Nickromancer** [comment]
Hello again @philipportner!
I've fixed the if statement now and added the ```PrintIRPass``` which was super useful, but my ```CreateDelilahContextOp``` should be called when creating the DelilahContext through the ```--delilah``` flag and not as a method in .daph. It isn't behaving properly even though I've followed the same steps as for ```CreateFPGAContextOp```, which makes me question if a CMAKE file is needed just like for FPGA. This is my error message:
```
dawn@nhed-2:~/daphne_dawn$ bin/daphne --delilah scripts/examples/Delilah.daph 
IR after Delilah Init:
module {
  func.func @main() {
    %0 = "daphne.constant"() {value = "/dev/delilah0"} : () -> !daphne.String
    "daphne.delilahInit"(%0) : (!daphne.String) -> ()
    "daphne.return"() : () -> ()
  }
}
JIT session error: Symbols not found: [ _createDelilahContext ]
JIT-Engine invocation failed: Failed to materialize symbols: { (main, { _mlir__mlir_ciface_main, _mlir_main, _mlir_ciface_main, main }) }Program aborted due to an unhandled Error:
Failed to materialize symbols: { (main, { _mlir__mlir_ciface_main, _mlir_main, _mlir_ciface_main, main }) }
bin/daphne(+0x14e50e2)[0x55b6a2cb40e2]
/lib/x86_64-linux-gnu/libc.so.6(+0x42520)[0x7f083da42520]
/lib/x86_64-linux-gnu/libc.so.6(pthread_kill+0x12c)[0x7f083da969fc]
/lib/x86_64-linux-gnu/libc.so.6(raise+0x16)[0x7f083da42476]
/lib/x86_64-linux-gnu/libc.so.6(abort+0xd3)[0x7f083da287f3]
bin/daphne(+0x3dc2dd0)[0x55b6a5591dd0]
bin/daphne(+0x1514243)[0x55b6a2ce3243]
bin/daphne(+0x151834d)[0x55b6a2ce734d]
/lib/x86_64-linux-gnu/libc.so.6(+0x29d90)[0x7f083da29d90]
/lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0x80)[0x7f083da29e40]
bin/daphne(+0x14e3be5)[0x55b6a2cb2be5]
[error]: Got an abort signal from the execution engine. Most likely an exception in a shared library. Check logs!
[error]: Execution error: Returning from signal 6
```
I have a deadline for finishing this integration between Daphne and Delilah on the 29th of April. I am incredibly grateful for your help so far and appreciate getting as far as possible before the end of the project. Thanks!

Working branch: https://github.com/Nickromancer/daphne_dawn_nick/tree/Delilah_init

**Comment by philipportner** [comment]
> Thanks for the fix @philipportner.
> 
> This seems to have caused a error with my `CreateDelilahContext.h`, where when build tells me that a method is a undefined symbol:
> 
> ```
> ld.lld: error: undefined symbol: DelilahContext::createDelilahContext(int)
> >>> referenced by kernels.cpp
> >>>               src/runtime/local/kernels/CMakeFiles/AllKernels.dir/kernels.cpp.o:(_createDelilahContext)
> collect2: error: ld returned 1 exit status
> ninja: build stopped: subcommand failed.
> ```
> 
> I looked at a different way of calling the method inspired by `CreateDistributedContext.h` but returns a different kind of error:
> 
> ```
> In file included from /home/dawn/daphne_dawn/build/src/runtime/local/kernels/kernels.cpp:19:
> /home/dawn/daphne_dawn/src/runtime/local/kernels/CreateDelilahContext.h: In function ‘void createDelilahContext(DaphneContext*)’:
> /home/dawn/daphne_dawn/src/runtime/local/kernels/CreateDelilahContext.h:29:71: error: no match for ‘operator=’ (operand types are ‘std::vector<std::unique_ptr<IContext> >’ and ‘std::unique_ptr<IContext>’)
>    29 |         ctx->delilah_contexts = DelilahContext::createDelilahContext(0);
>         |   
> ```
> 
> I'm unsure which direction is correct, thank you for your guidance! (Same branch as last comment) BR, Nicholas Hansen

This is very similar to before, you are again referring to a symbol that could not be resolved by the linker. 

Since you created a `DelilahContext.cpp` you need to make the compilation process aware of this translation unit, e.g., you need to add this to `src/runtime/local/kernels/CMakeLists.txt` to `SOURCES_cpp_kernels`.

**Comment by Nickromancer** [comment]
Hell again @philipportner!
I just found out that the DelilahContext isn't being built into ```main/..DaphneContext.h``` which is likely a issue.
I understand that I would maybe have to implement a ```CMAKE.txt``` for a folder that belongs to Delilah kernels and also make it built my ```DelilahContext.h``` just like ```FGPAOPENCL/CMakeLists.txt```, but what would be the bare minimum requirements to make it work while excluding FPGA specific parts of it?  

BR,
Nicholas Hansen
